### PR TITLE
perf(core): speed up promise hook dispatch

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: core/01_core.js [hunk 1]
```diff
@@ -326,43 +328,84 @@
...
+      // Add the material hook to the list.
+      ArrayPrototypePush(promiseHooks[i], hook);
+    }
 
-    if (!promiseHooks.hasBeenSet) {
-      promiseHooks.hasBeenSet = true;
+    const wrappedHooks = ArrayPrototypeMap(promiseHooks, (hooks) => {
+      switch (hooks.length) {
+        case 0:
+          return noop;
```
Comment (bartlomieju) [inline core/01_core.js:363]: We should make it `undefined` - with https://github.com/denoland/deno/pull/17604 landed we'll be able to not pass a callback for certain hooks

**Comment by bartlomieju** [comment]
Results on v1.30:

```
no hooks:
Iterations per second: 20746888
Iterations per second: 23201856
Iterations per second: 23310023

noop hooks:
Iterations per second: 9578544
Iterations per second: 9861933
Iterations per second: 9910803

AsyncLocalStorage:
Iterations per second: 2314815
Iterations per second: 2361833
Iterations per second: 2345766
```

This PR:

```
no hooks:
Iterations per second: 22075055
Iterations per second: 24570025
Iterations per second: 24630542

noop hooks:
Iterations per second: 10152284
Iterations per second: 8496177
Iterations per second: 10405827

AsyncLocalStorage:
Iterations per second: 2405581
Iterations per second: 2427774
Iterations per second: 2433682
```

Not much perf gain, but better than nothing

File: core/01_core.js [hunk 2]
```diff
@@ -326,43 +328,84 @@
...
+      // Verify that the type of `hook` is a function.
+      if (typeof hook !== "function") {
+        throw new TypeError(`Expected function at position ${i}`);
+      }
+      // Crudely filter out no-op functions.
+      if (
+        RegExpPrototypeTest(/\)(?:\s*|=>){}$/, FunctionPrototypeToString(hook))
+      ) {
+        continue;
+      }
```
Comment (bartlomieju) [inline core/01_core.js:355]: This is very iffy :/

File: core/01_core.js (hunk 1 above)
Comment (bartlomieju) [reply core/01_core.js:363]: Actually let's wait for https://github.com/denoland/deno/pull/17705

File: core/01_core.js (hunk 2 above)
Comment (piscisaureus) [reply core/01_core.js:355]: I agree, let me remove it

File: core/01_core.js [hunk 3]
```diff
@@ -326,43 +328,84 @@
...
+      // Add the material hook to the list.
```
Comment (piscisaureus) [inline core/01_core.js:356]:
```suggestion
      // Add the hook to the list.
```

**Comment by piscisaureus** [comment]
a bit disappointing indeed

File: core/01_core.js (hunk 1 above)
Comment (piscisaureus) [reply core/01_core.js:357]: yes i was anticipating that

Comment (piscisaureus) [reply core/01_core.js]: OK, done!

**Comment by piscisaureus** [comment]
With the latest version of this patch, I get the following benchmark results.
The benchmark installs two hooks (`init` and `before`) and performs minimalistic async context forwarding.

#### deno 1.30.3 (release, aarch64-apple-darwin)

```
piscisaureus@dumbo:~/Desktop/async_context 🦕 deno run bench.js
Iterations per second: 8992806
Iterations per second: 9157509
Iterations per second: 9208103
Iterations per second: 9523810
Iterations per second: 9478673
Iterations per second: 9551098
```

#### This PR

```
piscisaureus@dumbo:~/Desktop/async_context 🦕 ~/d/deno/target/release/deno run bench.js
Iterations per second: 10515247
Iterations per second: 10905125
Iterations per second: 10989011
Iterations per second: 11467890
Iterations per second: 11494253
Iterations per second: 11520737
```

**Pull request merged**